src/Features/Core/Portable/EditAndContinue/EditSession.cs

```diff
@@ -479,88 +576,260 @@ private async Task<ProjectChanges> GetProjectChangesAsync(Project project, Cance
+                    if (mvid == default)
+                    {
+                        if (mvidReadError != null)
+                        {
+                            diagnostics.Add((project.Id, ImmutableArray.Create(mvidReadError)));
+                        }
+
+                        EditAndContinueWorkspaceService.Log.Write("Emitting update of '{0}' [0x{1:X8}]: project not built", project.Id.DebugName, project.Id);
+                        continue;
+                    }
+
+                    var changedDocumentAnalyses = GetChangedDocumentsAnalyses(baseProject, project);
+                    var projectSummary = await GetProjectAnalysisSymmaryAsync(project, mvid, changedDocumentAnalyses, cancellationToken).ConfigureAwait(false);
+                    if (projectSummary != ProjectAnalysisSummary.ValidChanges)
+                    {
+                        Telemetry.LogProjectAnalysisSummary(projectSummary, ImmutableArray<Diagnostic>.Empty);
 
-                var pdbStream = new MemoryStream();
-                var updatedMethods = new List<MethodDefinitionHandle>();
+                        if (projectSummary == ProjectAnalysisSummary.CompilationErrors || projectSummary == ProjectAnalysisSummary.RudeEdits)
+                        {
+                            isBlocked = true;
+                        }
+
+                        continue;
+                    }
+
+                    var projectChanges = await GetProjectChangesAsync(changedDocumentAnalyses, cancellationToken).ConfigureAwait(false);
+                    var currentCompilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                    var allAddedSymbols = await GetAllAddedSymbolsAsync(project, cancellationToken).ConfigureAwait(false);
+                    var baseActiveStatements = await BaseActiveStatements.GetValueAsync(cancellationToken).ConfigureAwait(false);
+                    var baseActiveExceptionRegions = await BaseActiveExceptionRegions.GetValueAsync(cancellationToken).ConfigureAwait(false);
+                    var lineEdits = projectChanges.LineChanges.SelectAsArray((lineChange, p) => (p.GetDocument(lineChange.DocumentId).FilePath, lineChange.Changes), project);
+
+                    var emitter = new ProjectDeltaEmitter(baseProject, mvid, projectChanges, lineEdits, currentCompilation, allAddedSymbols, baseActiveStatements, baseActiveExceptionRegions, _nonRemappableRegions);
 
-                using (var metadataStream = SerializableBytes.CreateWritableStream())
-                using (var ilStream = SerializableBytes.CreateWritableStream())
+                    // dispatch to a background thread - the compiler reads symbols and ISymUnmanagedReader requires MTA thread:
+                    if (Thread.CurrentThread.GetApartmentState() != ApartmentState.MTA)
+                    {
+                        await Task.Factory.SafeStartNew(() => { }, cancellationToken, TaskScheduler.Default).ConfigureAwait(true);
```

Comment:
❗️ This should use the JTF form:

```suggestion
                        await TaskScheduler.Default;
```

This operation is a NOP when the code is already running on the specified scheduler, so the surrounding `if` check is unnecessary as well.

Reply:
The Features layer does not have a reference to VS threading. I think that's actually fine, since this code is gonna run on server.

---
In reply to: [270627000](https://github.com/dotnet/roslyn/pull/34027#discussion_r270627000) [](ancestors = 270627000)